### PR TITLE
[Frontend]-config-cli-args

### LINKED ISSUE #7350: [Feature]: Why vllm cli not provide a config arg?
### 🚀 The feature, motivation and pitch

for examaple:
  vllm serve --config config.json

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by KaunilD**
FILL IN THE PR DESCRIPTION HERE

FIX #7350 
Alows you to submit args using a config file. 

```shell
$ vllm serve --config <path/to/config.yaml>
```

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by DarkLight1337** [comment]
Can you add an example to the docs on using a config file?

File: vllm/scripts.py [hunk 1]
```diff
@@ -125,6 +133,13 @@ def main():
     serve_parser.add_argument("model_tag",
                               type=str,
                               help="The model tag to serve")
+    serve_parser.add_argument("config",
```
Comment (DarkLight1337) [inline vllm/scripts.py:136]:
```suggestion
    serve_parser.add_argument("--config",
```

I think it's supposed to be an option instead of a positional argument?

**Comment by KaunilD** [comment]
> Thanks for implementing this! Can you add to the docs an example of using a config file?

absolutely. on it.. just finished setting up my ubuntu box for e2e testing..

File: vllm/scripts.py (hunk 1 above)
Comment (KaunilD) [reply vllm/scripts.py]: good catch! thank you..

File: docs/source/serving/openai_compatible_server.md [hunk 2]
```diff
@@ -111,6 +111,29 @@ directory [here](https://github.com/vllm-project/vllm/tree/main/examples/)
...
+```yaml
+# config.yaml
+
+host: "127.0.0.1"
+port: 6379
+uvicorn-log-level: "info"
+```
+
+```bash
+$ vllm serve --model SOME_MODEL --config config.yaml
```
Comment (youkaichao) [inline docs/source/serving/openai_compatible_server.md:127]: `--model SOME_MODEL` can be in the `config.yaml` as well?

File: vllm/scripts.py [hunk 3]
```diff
@@ -25,6 +26,13 @@ def signal_handler(sig, frame):
 
 
 def serve(args: argparse.Namespace) -> None:
+    if args.config:
+        config_parser = configparser.ConfigParser()
+        config_parser.read(args.config)
+
+        for key, value in config_parser.items():
+            setattr(args, key, value)
+
```
Comment (youkaichao) [inline vllm/scripts.py:35]: if we have `--config`, make sure we don't have any other commandline args?

File: docs/source/serving/openai_compatible_server.md [hunk 4]
```diff
@@ -111,6 +111,29 @@ directory [here](https://github.com/vllm-project/vllm/tree/main/examples/)
...
+```
+
+---
+**NOTE** 
+
+In case an argument is supplied using CLI and the config file, the value from the config file will take precedence.
```
Comment (KaunilD) [inline docs/source/serving/openai_compatible_server.md:133]: NTS: delete this since cli and config args are mutually exclusive.

File: vllm/scripts.py (hunk 3 above)
Comment (youkaichao) [reply vllm/scripts.py:35]: wait, after a second though, it is not clear to me if we should make it mutually exclusive

Comment (youkaichao) [reply vllm/scripts.py:35]: for example. if I have a config that has lots of options, I might want to simply reuse it and override some values from the commandline, e.g. `vllm serve --config config.yaml -tp 2`

Comment (KaunilD) [reply vllm/scripts.py:35]: @youkaichao  howa bout this?

lets not make it exclusive to allow flexibility, BUT also raise a hard interrupt in case an option is specified in both cli and config files. this will prevent users from unknowingly launching their server with stale configs/args...

Comment (youkaichao) [reply vllm/scripts.py:35]: or print an INFO level logging, telling users explicitly that some arg value from commandline takes precedence ?

Comment (KaunilD) [reply vllm/scripts.py:35]: deal 🤝

File: docs/source/serving/openai_compatible_server.md (hunk 2 above)
Comment (KaunilD) [reply docs/source/serving/openai_compatible_server.md]: done. edited the doc.

File: vllm/scripts.py (hunk 3 above)
Comment (KaunilD) [reply vllm/scripts.py]: done. made cli the override

File: vllm/scripts.py [hunk 5]
```diff
@@ -24,7 +28,49 @@ def signal_handler(sig, frame):
...
+    try:
+        with open(args.config, 'r') as config_file:
+            config = yaml.safe_load(config_file)
+    except Exception as ex:
+        logger.error("Unable to read the config file at %s", args.config)
+        logger.error(ex)
+
+    for key, value in config.items():
+        if hasattr(args, key):
```
Comment (youkaichao) [inline vllm/scripts.py:59]: how does this deal with options that have default values?

Comment (KaunilD) [reply vllm/scripts.py:59]: tested. same behaviour.  default args are populated in the args.Namespace before it enters the method.

Comment (KaunilD) [reply vllm/scripts.py:59]: *same ebhaviour as in cli takes presidence, if specified in config file as well. and default in case no where speficied..

Comment (youkaichao) [reply vllm/scripts.py:59]: I mean, even if I don't specify `-tp 1`, `args.tensor_parallel` will have value. in this case, the default cli value takes precedence over config file, which is not desirable.

Comment (KaunilD) [reply vllm/scripts.py:59]: please bear with me @youkaichao 

i see so if i understand correctly.. ideal behavior would be that; the user supplies ALLthe args in config file, if --config has been used, even the default ones? In a manner of speaking the config file will always have ~105 lines in them (current population of namespace)..
so for example:

Ex1

```yaml
#config.yaml

port: 1234
```
```python
$ vllm serve "fabebook/opt-12b" --config config.yaml
$ [ERROR] Please specify all args in config file while using --config 
```

Ex2

```yaml
#config.yaml

port: 1234
... 104 more..
```
```python
$ vllm serve "fabebook/opt-12b" --config config.yaml --port 4
$ [INFO] port = 4
```

Ex3

```python
$ vllm serve "fabebook/opt-12b" 
$ [INFO] port = 8000 (default)
```

Comment (youkaichao) [reply vllm/scripts.py:59]: Looks like a misunderstanding.

I'm trying to understand, how do you know if `getattr(args, key)` is a default value or a value from commandline?

Comment (KaunilD) [reply vllm/scripts.py:59]: oooh got it! 
I understand it now. we will have to handle other cases as well with validating if arguments in config file have membership with the vllm namespace, shorthand etc and c.. 
with this in mind i think its prudent for us to use 1configargparse1 which inherits from 1argparse1.
this will keep the diff small with largest ROI as well. moreover, it doesnt neeed anything more than what is already in vllm's requirements.

File: docs/source/serving/openai_compatible_server.md [hunk 6]
```diff
@@ -111,6 +111,29 @@ directory [here](https://github.com/vllm-project/vllm/tree/main/examples/)
...
+$ vllm serve SOME_MODEL --config config.yaml
+```
+
+---
+**NOTE** 
+
+In case an argument is supplied using CLI and the config file, the value from the CLI will take precedence.
+
+---
+
```
[the remark this answers is not in the document]
Comment (KaunilD) [reply docs/source/serving/openai_compatible_server.md]: @DarkLight1337 unfortunately this is the markdown file so no support for sphinx/rust like `.. note::` annotation.

File: vllm/scripts.py (hunk 5 above)
Comment (KaunilD) [reply vllm/scripts.py]: PS:
I have also tested the the scratch implementation using sys.argv check for default membership and then overriding args appropirately. we can go ahead with that too if its too much of a leap in faith to use configargparse module..

File: docs/source/serving/openai_compatible_server.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply docs/source/serving/openai_compatible_server.md]: Didn't notice that, nevermind then

File: vllm/scripts.py (hunk 5 above)
Comment (KaunilD) [reply vllm/scripts.py]: @youkaichao  hope you are doing well. 

Unfortunately with `configargparse` the tests are not passing. The server is not able to start quickly enough for the api tests to pass..

I will go ahead with the sys.argv implementation since its "lightweight" and more tailor made for our purposes.

Comment (KaunilD) [reply vllm/scripts.py]: @youkaichao  i feel good  about this. Please let me know what do you think.

**Comment by KaunilD** [comment]
Hi @DarkLight1337  and @youkaichao  a gentle nudge on this one :)

**Comment by DarkLight1337** [comment]
I'll take another look at this later today.

**Comment by youkaichao** [comment]
sorry I don't have bandwidth to review this pr. @DarkLight1337  please go ahead to take the review.

My only ask is to keep the precedence level:

commandline args > config file > default value

e.g.

config.yaml:

```yaml
tp: 4
```

commandline:
`vllm server -tp 8`

and the result should be `tp=8` .

File: vllm/utils.py [hunk 7]
```diff
@@ -1108,6 +1112,104 @@ def parse_args(self, args=None, namespace=None):
...
+                "facebook/opt-12B", 
+                '--port', '12323', 
+                '--tensor-parallel-size', '4', 
+                '-tp', '2'
+                ]
+            ```
+
+            Please note how the config args are inserted after the sub command.
+            this way the order of priorities is maintained when these are args 
+            parsed by super().
```
Comment (DarkLight1337) [inline vllm/utils.py:1149]: There seems to be extra indentation here. The content should have the same indentation as the triple quotes.

File: vllm/utils.py [hunk 8]
```diff
@@ -1108,6 +1112,104 @@ def parse_args(self, args=None, namespace=None):
...
+        if index == len(args) - 1:
+            raise ValueError("No config file specified! \
+                             Please check your commandline arguments.")
+
+        file_path = args[index + 1]
+        extension: str = file_path.split('.')[-1]
+        if extension not in {'yaml', 'yml'}:
+            raise ValueError(
+                "Config file must be of a yaml/yml type.\
+                              %s supplied", extension)
```
Comment (DarkLight1337) [inline vllm/utils.py:1164]: Imo this part should be inside `_load_config_file`.

File: vllm/utils.py [hunk 9]
```diff
@@ -25,6 +25,7 @@
 import psutil
 import torch
 import torch.types
+import yaml
```
Comment (DarkLight1337) [inline vllm/utils.py:28]: Since this isn't in the standard library, we should add it to `requirements-common.txt`.

**Review by DarkLight1337 — COMMENTED** [review]
Left some comments!

File: docs/source/serving/openai_compatible_server.md [hunk 10]
```diff
@@ -111,6 +111,32 @@ directory [here](https://github.com/vllm-project/vllm/tree/main/examples/)
...
+port: 6379
+uvicorn-log-level: "info"
+```
+
+```bash
+$ vllm serve SOME_MODEL --config config.yaml
+```
+---
+**NOTE**  
+In case an argument is supplied using commandline and the config file, the value from the commandline will take precedence.
```
Comment (DarkLight1337) [inline docs/source/serving/openai_compatible_server.md:135]:
```suggestion
In case an argument is supplied using command line and the config file, the value from the commandline will take precedence.
```

File: vllm/utils.py [hunk 11]
```diff
@@ -1108,6 +1112,104 @@ def parse_args(self, args=None, namespace=None):
 
         return super().parse_args(processed_args, namespace)
 
+    @staticmethod
+    def _pull_args_from_config(args: List[str]) -> List[str]:
+        """
+            method to pull arguments specified in the config file
+            into the commandline args variable.
```
Comment (DarkLight1337) [inline vllm/utils.py:1119]:
```suggestion
            into the command-line args variable.
```

File: vllm/utils.py [hunk 12]
```diff
@@ -1108,6 +1112,104 @@ def parse_args(self, args=None, namespace=None):
...
+            this way the order of priorities is maintained when these are args 
+            parsed by super().
+        """
+        assert args.count(
+            '--config') <= 1, "More than one config file specified!"
+
+        index = args.index('--config')
+        if index == len(args) - 1:
+            raise ValueError("No config file specified! \
+                             Please check your commandline arguments.")
```
Comment (DarkLight1337) [inline vllm/utils.py:1157]:
```suggestion
                             Please check your command-line arguments.")
```

File: vllm/utils.py (hunk 9 above)
Comment (KaunilD) [reply vllm/utils.py:28]: good catch!

File: vllm/utils.py (hunk 7 above)
Comment (KaunilD) [reply vllm/utils.py]: changed and noted. same pattern with `_load_confug_file` as well.

File: vllm/utils.py (hunk 8 above)
Comment (KaunilD) [reply vllm/utils.py:1164]: done.

**Review by DarkLight1337 — APPROVED** [review]
LGTM now. Thanks for implementing this!
